### PR TITLE
[consensus] Implement Raft proxying PR 3/n

### PR DESCRIPTION
**Pull request by mpercy**
Backport of original patch written against the upstream Kudu master branch

This is a 3rd stack in a series after PR #42. Integration tests have been stripped since they are not currently supported in kuduraft.

The following is included in this stack:

- Support for tracking proxied messages and sending messages through proxies at the leader level
- Various cleanup and metrics

### CODE REVIEW

**Review by anirbanr-fb — APPROVED** [review]
Looked at this yesterday night. Approving 4/n

**Comment by mpercy** [comment]
@anirbanr-fb Thanks. Please let me know when you finish the stack, or at least get through 5/n. Patch 4/n is too minor to hassle splitting from this stack.

File: src/kudu/consensus/peer_manager.h [hunk 1]
```diff
@@ -71,14 +70,14 @@ class PeerManager {
  private:
   std::string GetLogPrefix() const;
 
-  typedef std::unordered_map<std::string, std::shared_ptr<Peer>> PeersMap;
```
Comment (anirbanr-fb) [inline src/kudu/consensus/peer_manager.h:74]: seems unnecessary change.

File: src/kudu/consensus/peer_manager.h [hunk 2]
```diff
@@ -71,14 +70,14 @@ class PeerManager {
...
   const std::string tablet_id_;
   const std::string local_uuid_;
   PeerProxyFactory* peer_proxy_factory_;
   PeerMessageQueue* queue_;
   ThreadPoolToken* raft_pool_token_;
   scoped_refptr<log::Log> log_;
-  PeersMap peers_;
+  PeerProxyPool peer_proxy_pool_;
+  std::unordered_map<std::string, std::shared_ptr<Peer>> peers_;
```
Comment (anirbanr-fb) [inline src/kudu/consensus/peer_manager.h:80]: PeersMap seems like a good idea.

File: src/kudu/consensus/consensus_queue.cc [hunk 3]
```diff
@@ -1025,6 +1050,10 @@ void PeerMessageQueue::EndWatchForSuccessor() {
   tl_filter_fn_ = nullptr;
 }
 
+Status PeerMessageQueue::GetNextRoutingHopFromLeader(const string& dest_uuid, string* next_hop) const {
+  return routing_table_->NextHop(local_peer_pb_.permanent_uuid(), dest_uuid, next_hop);
```
Comment (anirbanr-fb) [inline src/kudu/consensus/consensus_queue.cc:1054]: does this not need protection by lock? The lifetime of PeerMessageQueue should involve multiple changes to routing_table_?

File: src/kudu/consensus/consensus_queue.cc [hunk 4]
```diff
@@ -1281,6 +1310,10 @@ bool PeerMessageQueue::ResponseFromPeer(const std::string& peer_uuid,
   {
     std::lock_guard<simple_spinlock> scoped_lock(queue_lock_);
 
+    // TODO(mpercy): Handle response from proxy on behalf of another peer.
```
Comment (anirbanr-fb) [inline src/kudu/consensus/consensus_queue.cc:1313]: I am not sure I understand this comment.

**Review by anirbanr-fb — APPROVED** [review]
I will approve, but the routing table thread safety needs to be looked into

File: src/kudu/consensus/consensus_queue.cc (hunk 3 above)
Comment (mpercy) [reply src/kudu/consensus/consensus_queue.cc:1054]: No, routing_table_ is set only once (at construction time) and it's a shared_ptr to an instance of DurableRoutingTable, which is thread-safe. DRT lives forever (as long as ConsensusMetadataManager lives) or until it's deleted (in kuduraft it's never deleted, but that would only happen when destroying the tablet including the consensus instance).

File: src/kudu/consensus/consensus_queue.cc (hunk 4 above)
Comment (mpercy) [reply src/kudu/consensus/consensus_queue.cc:1313]: Right now we don't handle proxying in a special way when we receive the response at the leader. Maybe in the future we might want to report status in a special way if we know that the proxy is down as opposed to the destination node being down. I wasn't sure if it would help operability to add that so for now I didn't, I simply added a TODO.

File: src/kudu/consensus/peer_manager.h (hunk 1 above)
Comment (mpercy) [reply src/kudu/consensus/peer_manager.h:74]: This was previously used at some point in the rest of the code however having this typedef for a single definition (in the same header file) seemed superfluous. However you're right, this change is basically a non-sequitur.

File: src/kudu/consensus/peer_manager.h (hunk 2 above)
Comment (mpercy) [reply src/kudu/consensus/peer_manager.h:80]: See above comment

**Comment by mpercy** [comment]
> I will approve, but the routing table thread safety needs to be looked into

Thanks for the review! 6 & 7 should be an easy stamp, LMK if you have any questions!

File: src/kudu/consensus/routing.cc [hunk 5]
```diff
@@ -274,18 +274,24 @@ Status DurableRoutingTable::Create(FsManager* fs_manager,
...
   ProxyTopologyPB proxy_topology;
-  RETURN_NOT_OK(pb_util::ReadPBContainerFromPath(fs_manager->env(),
-                                                 path,
-                                                 &proxy_topology));
-
-  *drt = std::shared_ptr<DurableRoutingTable>(new DurableRoutingTable(fs_manager,
-                                          std::move(tablet_id),
-                                          std::move(proxy_topology),
-                                          std::move(raft_config)));
+  Status s = pb_util::ReadPBContainerFromPath(fs_manager->env(),
```
Comment (anirbanr-fb) [inline src/kudu/consensus/routing.cc:282]: need a conversation with ritwik on how to persist critical information which is beyond standard "raft configuration"

**Review by anirbanr-fb — APPROVED** [review]

**Review by anirbanr-fb — APPROVED** [review]
cursory look only, but seems like valuable metrics.

**Comment by mpercy** [comment]
Thanks for the review! I'll merge this now.
